File: saumon_fumé/lib/env.ml
type t =
  | Global of Value.t option array
  | Local of Value.t array * t

let global r =
  let open Resolver in
  let frame = Array.make (slots r.global_frame) None in
  Global frame

(* TODO: make this safer somehow? *)
let expand e globals = match e with
  | Local _ -> e
  | Global frame ->
      let global_slots = Resolver.slots globals in
      let len = Array.length frame in
      let gap = global_slots - len in
      if gap > 0
        then
          let extra = Array.make gap None in
          Global (Array.append frame extra)
        else e

let rec read e var = match e, var with
  | Local (frame, _), (_, 0, slot) ->
      Some frame.(slot)
  | Local (_, parent), (name, n, slot) when n > 0 ->
      read parent (name, n - 1, slot)
  | Global frame, (_, 0, slot) ->
      frame.(slot)
  | _, (item, _, _) ->
      failwith ("internal error: bad lookup for " ^ item)

let rec define e var v = match e, var with
  | Local (frame, _), (_, 0, slot) ->
      frame.(slot) <- v
  | Local (_, parent), (name, n, slot) when n > 0 ->
      define parent (name, n - 1, slot) v
  | Global frame, (_, 0, slot) ->
      frame.(slot) <- Some v
  | _, (item, _, _) ->
      failwith ("internal error: bad lookup for " ^ item)

let rec assign e var v = match e, var with
  | Local (frame, _), (_, 0, slot) ->
      frame.(slot) <- v;
      true
  | Local (_, parent), (name, n, slot) when n > 0 ->
      assign parent (name, n - 1, slot) v
  | Global frame, (_, 0, slot) ->
      begin match frame.(slot) with
        | Some _ ->
            frame.(slot) <- Some v;
            true
        | None -> false
      end
  | _, (item, _, _) ->
      failwith ("internal error: bad lookup for " ^ item)

let push e slots = Local (Array.make slots Value.Nil, e)

let pop = function
  | Global _ -> failwith "internal error: pop from global env"
  | Local (_, parent) -> parent

File: loxide/src/env.rs
use std::{
    cell::{Ref, RefMut, RefCell,},
    rc::Rc,
};

use crate::{
    value::Value,
};

#[derive(Clone, Debug)]
pub struct Slot {
    pub name: String,
    pub frame: usize,
    pub index: usize,
}

// this will all change
#[derive(Clone, Debug)]
pub enum Env {
    Global(RefCell<Vec<Option<Value>>>),
    Local(RefCell<Vec<Value>>, Rc<Env>),
}

impl Env {
    #[inline]
    pub fn new(slots: usize) -> Rc<Self> {
        Rc::new(Env::Global(RefCell::new(vec![None; slots])))
    }

    #[inline]
    pub fn child(self: &Rc<Self>, slots: usize) -> Rc<Self> {
        Rc::new(
          Env::Local(RefCell::new(vec![Value::Nil; slots]), Rc::clone(self)))
    }

    pub fn get(&self, frame: usize, index: usize) -> Option<Ref<Value>> {
        if frame == 0 {
            match self {
                Env::Global(slots) =>
                    Ref::filter_map(slots.borrow(), |v| v[index].as_ref()).ok(),
                Env::Local(slots, _) =>
                    Some(Ref::map(slots.borrow(), |v| &v[index]))
            }
        } else {
            match self {
                Env::Global(_) => panic!("internal error: bad lookup"),
                Env::Local(_, parent) => parent.get(frame - 1, index),
            }
        }
    }

    pub fn get_mut(&self, frame: usize, index: usize) -> Option<RefMut<Value>> {
        if frame == 0 {
            match self {
                Env::Global(slots) =>
                    RefMut::filter_map(slots.borrow_mut(),
                      |v| v[index].as_mut()).ok(),
                Env::Local(slots, _) =>
                    Some(RefMut::map(slots.borrow_mut(), |v| &mut v[index]))
            }
        } else {
            match self {
                Env::Global(_) => panic!("internal error: bad lookup"),
                Env::Local(_, parent) => parent.get_mut(frame - 1, index),
            }
        }
    }

    pub fn set(&self, frame: usize, index: usize, value: Value) {
        if frame == 0 {
            match self {
                Env::Global(slots) =>
                    slots.borrow_mut()[index] = Some(value),
                Env::Local(slots, _) =>
                    slots.borrow_mut()[index] = value,
            }
        } else {
            match self {
                Env::Global(_) => panic!("internal error: bad lookup"),
                Env::Local(_, parent) => parent.set(frame - 1, index, value),
            }
        };
    }

    #[inline]
    pub fn ensure_slots(&self, slots: usize) {
        match self {
            Env::Global(v) => {
                let mut v = v.borrow_mut();
                while v.len() < slots {
                    v.push(None);
                }
            },
            Env::Local(v, _) => {
                let mut v = v.borrow_mut();
                while v.len() < slots {
                    v.push(Value::Nil);
                }
            },
        }
    }
}
